src/main/java/org/apache/bytes/Bytes.java

```diff
@@ -0,0 +1,452 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.bytes;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.lang.ref.WeakReference;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Represents bytes in Fluo. Bytes is an immutable wrapper around a byte array. Bytes always copies
+ * on creation and never lets its internal byte array escape. Its modeled after Java's String which
+ * is an immutable wrapper around a char array. It was created because there is nothing in Java like
+ * it at the moment. Its very nice having this immutable type, it avoids having to do defensive
+ * copies to ensure correctness. Maybe one day Java will have equivalents of String, StringBuilder,
+ * and Charsequence for bytes.
+ *
+ * <p>
+ * The reason Fluo did not use ByteBuffer is because its not immutable, even a read only ByteBuffer
+ * has a mutable position. This makes ByteBuffer unsuitable for place where an immutable data type
+ * is desirable, like a key for a map.
+ *
+ * <p>
+ * Bytes.EMPTY is used to represent a Bytes object with no data.
+ *
+ * @since 1.0.0
+ */
+public final class Bytes extends AbstractByteSequence implements Comparable<Bytes>, Serializable {
+
+  private static final long serialVersionUID = 1L;
+
+  private final byte[] data;
+
+  private transient WeakReference<String> utf8String;
+
+  public static final Bytes EMPTY = new Bytes(new byte[0]);
+
+  private int hashCode = 0;
+
+  public Bytes() {
+    data = EMPTY.data;
+  }
+
+  private Bytes(byte[] data) {
+    this.data = data;
+  }
+
+  private Bytes(byte[] data, String utf8String) {
+    this.data = data;
+    this.utf8String = new WeakReference<>(utf8String);
+  }
+
+  /**
+   * Gets a byte within this sequence of bytes
+   *
+   * @param i index into sequence
+   * @return byte
+   * @throws IllegalArgumentException if i is out of range
+   */
+  @Override
+  public byte byteAt(int i) {
+    checkBounds(i);
+    return data[i];
+  }
+
+  /**
+   * Gets the length of bytes
+   */
+  @Override
+  public int length() {
+    return data.length;
+  }
+
+  /**
+   * Returns a portion of the Bytes object
+   *
+   * @param begin index of subsequence begin (inclusive)
+   * @param end index of subsequence end (exclusive)
+   */
+  @Override
+  public Bytes subSequence(int begin, int end) {
+    checkBounds(begin, end);
+    return Bytes.of(data, begin, end - begin);
+  }
+
+  /**
+   * Returns a byte array containing a copy of the bytes
+   */
+  public byte[] toArray() {
```

Comment:
Wondering if this method should be in ByteSequence.

Reply:
Oh, yeah, that makes sense.